src/lib.rs

```diff
@@ -1,14 +1,165 @@
+                next_level.push(parent_hash);
+            }
+            current_level = next_level;
+        }
+        Some(current_level[0].clone())
+    }
+
+    /// Adds a new leaf element and builds the Merkle Tree again.
+    ///
+    /// # Arguments
+    ///
+    /// * `element` - A string slice representing the new element to be added.
+    pub fn add_element(&mut self, element: &str) {
+        let new_leaf = Self::hash(element);
+        self.leaves.push(new_leaf.clone());
+        self.root = Self::build_tree(&self.leaves);
+    }
+
+    /// Generates a proof that an element is contained in the Merkle Tree.
+    ///
+    /// # Arguments
+    ///
+    /// * `element` - A string slice representing the element for which the proof is to be generated.
+    ///
+    /// # Returns
+    ///
+    /// An `Option<Vec<(String, bool)>>` representing the proof, where each tuple contains a hash
+    /// and a boolean indicating whether the hash is from the left (true) or right (false) sibling.
+    pub fn generate_proof(&self, element: &str) -> Option<Vec<(String, bool)>> {
+        if self.leaves.is_empty() {
+            return None;
+        }
+
+        let mut hash = Self::hash(element);
+        let mut proof = Vec::new();
+        let mut current_level = self.leaves.clone();
+
+        while current_level.len() > 1 {
+            let mut next_level = Vec::new();
+            let mut found = false;
+
+            for i in (0..current_level.len()).step_by(2) {
+                let left = &current_level[i];
+                let right = if i + 1 < current_level.len() {
+                    &current_level[i + 1]
+                } else {
+                    left
+                };
+                let parent_hash = Self::hash(&(left.clone() + right));
+                if left == &hash || right == &hash {
+                    proof.push(if left == &hash {
+                        (right.clone(), false)
+                    } else {
+                        (left.clone(), true)
+                    });
+                    hash.clone_from(&parent_hash);
+                    found = true;
+                }
+                next_level.push(parent_hash);
+            }
+
+            if !found {
+                return None;
+            }
+
+            current_level = next_level;
+        }
+
+        Some(proof)
+    }
+
+    /// Verifies that an element is part of the Merkle Tree using the given proof.
+    ///
+    /// # Arguments
+    ///
+    /// * `element` - A string slice representing the element to be verified.
+    /// * `proof` - A vector of tuples representing the proof, where each tuple contains a hash
+    ///   and a boolean indicating whether the hash is from the left (true) or right (false) sibling.
+    ///
+    /// # Returns
+    ///
+    /// A boolean indicating whether the element is part of the tree.
+    pub fn verify(&self, element: &str, proof: Vec<(String, bool)>) -> bool {
```

Comment:
Maybe this could be improved using a `Result` with a custom error? I don't think the booleans are wrong, I just feel the Rust approach is nice.

Reply:
I'm not sure about this one, do you think that returning a `Result<(), CustomError>` is better than returning a `bool`? Or you meant `Result<bool, CustomError>`?

Reply:
I meant `Result<(), CustomError>`, where `Ok(())` would be the `true` case and `Err(CustomError)` would be the `false` case (unverified). I think either `bool` and `Result` are okay, it's fine if you prefer to keep the `bool`.